src/Dialogs/Frequency/dlgUserFrequencyList.cpp

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The XCSoar Project
+
+#include "dlgUserFrequencyList.hpp"
+#include "LogFile.hpp"
+#include "system/Path.hpp"
+#include "io/FileOutputStream.hxx"
+#include "io/BufferedCsvReader.hpp"
+#include "io/FileReader.hxx"
+#include "io/BufferedOutputStream.hxx"
+#include "io/BufferedReader.hxx"
+#include "LocalPath.hpp"
+#include "io/DataFile.hpp"
+#include "io/StringConverter.hpp"
+#include "ActionInterface.hpp"
+#include "UIGlobals.hpp"
+#include "dlgFrequencyEdit.hpp"
+#include "Dialogs/Message.hpp"
+
+void UserFrequencyListWidget::ReadFile()
+{
+  list.clear();
+  try
+  {
+    auto reader = OpenDataFile(file_name);
+    BufferedReader buffered_reader{*reader};
+    StringConverter string_converter(Charset::UTF8);
+
+    std::array<std::string_view, 4> params;
+
+    while (ReadCsvRecord(buffered_reader, params))
+    {
+      RadioFrequency freq = RadioFrequency::Parse(params.at(2));
+      if (freq.IsDefined() == false)
+        continue;
+
+      tstring_view name = string_converter.Convert(params.at(0));
+      tstring_view short_name = string_converter.Convert(params.at(1));
+      list.emplace_back(name, short_name, freq);
+    }
+  }
+  catch (...)
+  {
+  }
+
+  listUpdated();
+}
+
+void UserFrequencyListWidget::listUpdated()
+{
+  if (edit_mode)
+  {
+    edit_button->SetEnabled(list.size() > 0);
+    delete_button->SetEnabled(list.size() > 0);
+    duplicate_button->SetEnabled(list.size() > 0);
+    up_button->SetEnabled(list.size() > 1);
+    down_button->SetEnabled(list.size() > 1);
+  }
+  else
+  {
+    set_button->SetEnabled(list.size() > 0);
+    set_standby_button->SetEnabled(list.size() > 0);
+  }
+
+  ListControl &list_control = GetList();
+  list_control.SetLength(list.size());
+  list_control.Invalidate();
+}
+
+void UserFrequencyListWidget::UpdateFile()
+{
+  const auto path = LocalPath(file_name);
+
+  FileOutputStream file(path);
+  BufferedOutputStream writer(file);
+
+  for (const auto &item : list)
+  {
+    StaticString<8> freq_str;
+    item.freq.Format(freq_str.buffer(), freq_str.capacity());
+
+    writer.Write(item.name);
+    writer.Write(_T(","));
+    writer.Write(item.short_name);
+    writer.Write(_T(","));
+    writer.Write(freq_str);
+    writer.Write(_T("\n"));
+  }
+
+  writer.Flush();
+  file.Commit();
+
+  LogFormat(_T("Frequencies file '%s' saved"), path.c_str());
+}
+
+void UserFrequencyListWidget::CreateButtons(WidgetDialog &dialog)
+{
+  form = &dialog;
+  if (!edit_mode)
+  {
+    set_button = dialog.AddButton(_("Set Active"), [this]()
+                                  { SetClicked(true); });
+    set_standby_button = dialog.AddButton(_("Set Stand-by"), [this]()
+                                          { SetClicked(false); });
+    edit_mode_button = dialog.AddButton(_("Edit"), [this]()
```

Comment:
I think edit here is imprecise as it would lead the user to believe that this edits the current frequency. Please use "Manage" for this stage, Then edit for editing the items.